jupyterhub/auth.py

```diff
@@ -643,25 +666,50 @@ def authenticate(self, handler, data):
         Return None otherwise.
         """
         username = data['username']
+        admin = None
         try:
             pamela.authenticate(username, data['password'], service=self.service, encoding=self.encoding)
         except pamela.PAMError as e:
             if handler is not None:
                 self.log.warning("PAM Authentication failed (%s@%s): %s", username, handler.request.remote_ip, e)
             else:
                 self.log.warning("PAM Authentication failed: %s", e)
-        else:
-            if not self.check_account:
-                return username
+            return None
+    
+        if self.check_account:
             try:
                 pamela.check_account(username, service=self.service, encoding=self.encoding)
             except pamela.PAMError as e:
                 if handler is not None:
                     self.log.warning("PAM Account Check failed (%s@%s): %s", username, handler.request.remote_ip, e)
                 else:
                     self.log.warning("PAM Account Check failed: %s", e)
-            else:
-                return username
+                return None
+
+        if username in self.admin_users:
+            admin = True
+        elif self.admin_groups:
+            try:
+                # Most likely source of error here is a group name <-> gid mapping failure
```

Comment:
I'd rather fail on errors here.

In fact, would you like to refactor this step into a `def is_admin(username): bool|None` method? That would make it easier to override just this logic in subclasses for those who want to have different behavior on errors.

Reply:
That sounds wisest.

I suppose `is_admin` should go in `Authenticator`, do you agree? I'll start moving things around!